### ISSUE TITLE
Generic return type is not deduced even if a variable of the same type is.

### ISSUE THREAD

**Issue by Denaun** [microsoft/pyright#1295]
Moving a discussion from #1278.
I am seeing complaints from `pylance` with generic return types when they are deduced inside a function (e.g., through `isinstance` checks).

# Minimum reproducible example

```python
from typing import TypeVar

T = TypeVar('T')


def do_nothing(x: T) -> T:
    if isinstance(x, int):
        return x
    return x
```

Here `mypy` agrees with `pylance` saying that, on line 8, `x` (of type `int`) cannot be assigned to `T`.
While this might be the intended behaviour, my question is: why can `T` be deduced to be `int` when calling the function (e.g., `do_nothing(1)` has return type `int`) but not from within the function?

# With constrains

```python
from typing import TypeVar

T = TypeVar('T', int, str)  # Only difference with the code above.


def do_nothing(x: T) -> T:
    if isinstance(x, int):
        return x
    return x
```

Here I think `pylance` is being smarter than `mypy`, being able to understand that `x` is of type `int` on line 8 and of type `str` on line 9. This leads `pylance` to yield two errors of the same kind as above. `mypy` seems happy with the code.

# With bounds

I am adding this example just because @erictraut mentioned constrained vs bounded generics. Apologies if I missed the point.

```python
from typing import TypeVar


class Bar:
    pass


class Foo(Bar):
    pass


T = TypeVar('T', bound=Bar)


def do_nothing(x: T) -> T:
    if isinstance(x, Foo):
        return x
    return x
```

The behaviour here is the same as the first example: `x` of type `Foo` can't be assigned to `T`. Both `mypy` and `pylance` complain about it. Again I don't see the reason for this since `do_nothing(Foo())` correctly has return type `Foo`.

Thank you,
Maurizio

**Comment by erictraut** [comment]
All of these are edge cases, and there's not necessarily a "right answer" for any of them. I think you've made a good case for the "isinstance" narrowing to handle the constrained TypeVar case in a way that preserves the ability to assign the narrowed expression back to the original TypeVar. I don't think it makes sense in the bound or unconstrained/unbound cases. So I've kept those cases as is, which is consistent with mypy. I've augmented the handling of constrained TypeVars as you've suggested.

This change will be in the next release.

**Comment by Denaun** [comment]
Sounds good, thank you!

**Comment by erictraut** [comment]
We'll leave the issue open until the fix is released.

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.96, which I just published. It will also be in the next published release of Pylance.
